From 592d0d318685e85fba998d3b259fd644dee6f340 Mon Sep 17 00:00:00 2001
From: Neil Horman <nhorman@openssl.org>
Date: Sat, 16 Dec 2023 15:32:48 -0500
Subject: [PATCH 014/228] Check for NULL cleanup function before using it in
 encoder_process

encoder_process assumes a cleanup function has been set in the currently
in-use encoder during processing, which can lead to segfaults if said
function hasn't been set

Add a NULL check for this condition, returning -1 if it is not set

Reviewed-by: Tomas Mraz <tomas@openssl.org>
Reviewed-by: Matt Caswell <matt@openssl.org>
(Merged from https://github.com/openssl/openssl/pull/23069)

(cherry picked from commit cf57c3ecfa416afbc47d36633981034809ee6792)
---
 crypto/encode_decode/encoder_lib.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/crypto/encode_decode/encoder_lib.c b/crypto/encode_decode/encoder_lib.c
index 7868da79b7..9a0c689ec0 100644
--- a/crypto/encode_decode/encoder_lib.c
+++ b/crypto/encode_decode/encoder_lib.c
@@ -59,6 +59,11 @@ int OSSL_ENCODER_to_bio(OSSL_ENCODER_CTX *ctx, BIO *out)
         return 0;
     }
 
+    if (ctx->cleanup == NULL || ctx->construct == NULL) {
+        ERR_raise(ERR_LIB_OSSL_ENCODER, ERR_R_INIT_FAIL);
+        return 0;
+    }
+
     return encoder_process(&data) > 0;
 }
 
-- 
2.39.5 (Apple Git-154)

